warehouse/events/models.py

```diff
@@ -129,6 +129,18 @@ def record_event(
     ):
         """Records an Event record on the associated model."""
         session = orm.object_session(self)
+
+        if request is not None:
+            # Add `request.ip_address` data to `Event.additional` as a nested key
+            additional = additional or {}
+            additional["ip_address"] = {}
```

Comment:
I'd remove these, and instead populate the `ip_address_obj` relation on the event if we have an `IPAddress`!

Something like

```
ip_address_obj = None
if request is not None:
    ip_address_obj = request.ip_address
    additional = additional or {}
    additional["geoip_info"] = (
        request.ip_address.geoip_info or None
    )

event = self.Event(
    source=self,
    tag=tag,
    ip_address=ip_address,
    ip_address_obj=ip_address_obj,
    additional=additional,
)

Reply:
We discussed in depth synchronously, and I think I got the gist of it.

A refactor commit was added - it's a bit more involved than I had hoped for, but I think it does the trick.